### PR TITLE
[Dev] Implement blocksparse attention

### PR DESCRIPTION
**Pull request by smallscientist1**
This PR includes support for block sparse attention, similar to the mask API in `FlexAttention`. 
For example,
```py
BLOCK = 128
def block_sparse_mask(b, h, q_idx, kv_idx):
    return torch.logical_and(q_idx >= kv_idx, q_idx//BLOCK < kv_idx//BLOCK + 2)

```

Future TODOs:
- [ ] Support for arbitrary sparse mask, which need mask within single block.

### CODE REVIEW

**Comment by smallscientist1** [comment]
@xiayuqing0622 Ready to merge.
